source/common/upstream/upstream_impl.cc

```diff
@@ -38,15 +39,45 @@ namespace {
 
 const Network::Address::InstanceConstSharedPtr
 getSourceAddress(const envoy::api::v2::Cluster& cluster,
-                 const Network::Address::InstanceConstSharedPtr source_address) {
+                 const envoy::api::v2::core::BindConfig& bind_config) {
   // The source address from cluster config takes precedence.
   if (cluster.upstream_bind_config().has_source_address()) {
     return Network::Address::resolveProtoSocketAddress(
         cluster.upstream_bind_config().source_address());
   }
   // If there's no source address in the cluster config, use any default from the bootstrap proto.
-  return source_address;
+  if (bind_config.has_source_address()) {
+    return Network::Address::resolveProtoSocketAddress(bind_config.source_address());
+  }
+
+  return nullptr;
+}
+
+uint64_t parseFeatures(const envoy::api::v2::Cluster& config,
+                       const envoy::api::v2::core::BindConfig bind_config) {
+  uint64_t features = 0;
+  if (config.has_http2_protocol_options()) {
+    features |= ClusterInfoImpl::Features::HTTP2;
+  }
+  if (config.protocol_selection() == envoy::api::v2::Cluster::USE_DOWNSTREAM_PROTOCOL) {
+    features |= ClusterInfoImpl::Features::USE_DOWNSTREAM_PROTOCOL;
+  }
+  // Cluster IP_FREEBIND settings, when set, will override the cluster manager wide settings.
+  if ((bind_config.freebind().value() && !config.upstream_bind_config().has_freebind()) ||
+      config.upstream_bind_config().freebind().value()) {
+    features |= ClusterInfoImpl::Features::FREEBIND;
+  }
+  return features;
 }
+
+// Socket::Option implementation for API-defined listener upstream options.
```

Comment:
What is a "listener upstream option?" Clarify?

Reply:
That's a copy-paste snafu, fixed.